File: SparkStreaming/src/main/java/com/projects/bigdata/spark_streaming/AbstractStreaming.java
package com.projects.bigdata.spark_streaming;

import org.apache.spark.SparkConf;
import org.apache.spark.streaming.Durations;
import org.apache.spark.streaming.api.java.JavaStreamingContext;

import com.projects.bigdata.spark_streaming.utility.StatefulAggregationType;
import com.projects.bigdata.spark_streaming.utility.StreamingAppParameters;

public abstract class AbstractStreaming {

    final String dataStreamHost1;
    final int dataStreamPort1;
    final String dataStreamHost2;
    final int dataStreamPort2;
    final long processingTimeout;
    final StatefulAggregationType statefulAggregationType;
    final JavaStreamingContext streamingContext;

    AbstractStreaming(final StreamingAppParameters sap) {
        this.dataStreamHost1 = sap.getDataStreamHost1();
        this.dataStreamPort1 = sap.getDataStreamPort1();
        this.dataStreamHost2 = sap.getDataStreamHost2();
        this.dataStreamPort2 = sap.getDataStreamPort2();
        this.processingTimeout = sap.getProcessingTimeout();
        this.statefulAggregationType = sap.getStatefulAggregationType();
        streamingContext = new JavaStreamingContext(new SparkConf().setMaster("local["+sap.getNrThreads()+"]").setAppName("StreamingWordCount"), Durations.seconds(sap.getBatchDuration()));
        streamingContext.checkpoint(sap.getCheckpointDir());
    }

    abstract void process() throws Exception;

    void startStreamingAndAwaitTerminationOrTimeout(final long timeOutMilliSeconds) throws InterruptedException {
        streamingContext.start();
        streamingContext.awaitTerminationOrTimeout(processingTimeout);
    }
}

File: SparkStreaming/src/main/kotlin/com/projects/bigdata/spark_streaming/AbstractStreaming.kt
package com.projects.bigdata.spark_streaming

import com.projects.bigdata.spark_streaming.utility.StreamingAppParameters
import org.apache.spark.SparkConf
import org.apache.spark.streaming.Durations
import org.apache.spark.streaming.api.java.JavaStreamingContext

abstract class AbstractStreaming(val sap: StreamingAppParameters) {

    internal val streamingContext: JavaStreamingContext = JavaStreamingContext(SparkConf().setMaster("local[" + sap.nrThreads + "]").setAppName("StreamingWordCount"), Durations.seconds(sap.batchDuration.toLong()))

    init {
        streamingContext.checkpoint(sap.checkpointDir)
    }

    internal abstract fun process()

    internal fun startStreamingAndAwaitTerminationOrTimeout() {
        with (streamingContext) {
            start()
            awaitTerminationOrTimeout(sap.processingTimeout)
        }
    }
}